### PR TITLE
Attach to a process via ptrace.

### PR DESCRIPTION
**Pull request by danielfenner**
Attach to and stop all threads of a process. Also detach again.

Test: Unit Test.

### CODE REVIEW

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+// Attaches to the thread using ptrace.
+// Returns true if the thread was halted and false if the thread did not exist anymore.
+[[nodiscard]] ErrorMessageOr<bool> AttachAndStopThread(pid_t tid) {
+  if (ptrace(PTRACE_ATTACH, tid, nullptr, nullptr) == -1) {
+    // If tid has ended already we get ESRCH; if the thread was in 'exit state' we get EPERM.
+    // There are a bunch of other (non-relevant) cases. I haven't found documentation on this but it
+    // can be looked up in the function 'ptrace_attach' in 'ptrace.c' in the kernel sources.
+    if (errno == ESRCH || errno == EPERM) {
+      return false;
+    } else {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:38]: drop else?

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      return ErrorMessage(
+          absl::StrFormat("PTRACE_ATTACH failed for \"%d\": \"%s\"", tid, strerror(errno)));
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:40]: Use SafeStrerror from orbit base (the glibc one is not guaranteed to be thread-safe)

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+  // Wait for the traced thread to stop. Timeout after one second.
+  int32_t timeout_ms = 1000;
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:44]: nitpick, kind of... these are more like iterations and we we alwasy get more than 1 second. I do not think we need accuarcy here so maybe rename it to max_attemts? In which case the while below can be a for loop.

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    if (result_waitpid == -1) {
+      return ErrorMessage(absl::StrFormat(
+          "Wait for thread to get traced failed for tid \"%d\": \"%s\"", tid, strerror(errno)));
+    }
+    if (result_waitpid > 0) {
+      // Occasionally the thread is active during PTRACE_ATTACH but terminates before it gets
+      // descheduled. So waitpid returns on exit of the thread instead of the expected stop.
+      if (WIFEXITED(stat_val)) {
+        return false;
+      } else if (WIFSTOPPED(stat_val)) {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:57]: nit: drop else?

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        return true;
+      } else {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:59]: and here.

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      }
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    if (--timeout_ms == 0) {
+      ptrace(PTRACE_DETACH, tid, nullptr, nullptr);
+      return ErrorMessage(
+          absl::StrFormat("Waiting for the traced thread \"%d\" to stop timed out.", tid));
+    }
+  }
+  return true;
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:72]: this line seems to be unreachable....

File: src/UserSpaceInstrumentation/AttachTest.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  TestProcess test_process;
+  const pid_t pid = test_process.pid();
+
+  const auto result = AttachAndStopProcess(pid);
+  ASSERT_TRUE(result) << result.error().message();
+
+  // TestProcess is continuously spawning new threads when it is running. Verify that no new threads
+  // get spawned - i.e. the process is not running anymore.
+  const auto tids = orbit_base::GetTidsOfProcess(pid);
+  absl::SleepFor(absl::Milliseconds(50));
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/AttachTest.cpp:28]: this is not related to this PR: I was thinking that maybe we should start moving away from absl time functions. that is if stl already has everything we need.

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  while (true) {
+    if (std::chrono::system_clock::now() > deadline) {
+      break;
+    }
+  }
+  absl::MutexLock lock{&joinable_threads_mutex_};
+  joinable_threads_.emplace(std::this_thread::get_id());
+}
+
+void TestProcess::DummyWorkload() {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.cpp:66]: Fake?

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+TestProcess::~TestProcess() {
+  fs::remove(flag_file_run_child_);
+  int status;
+  waitpid(pid_, &status, 0);
+  CHECK(WIFEXITED(status));
+  fs::remove(flag_file_child_started_);
+}
+
+void TestProcess::DummyWorker() {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.cpp:54]: Fake?

File: src/UserSpaceInstrumentation/TestProcess.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace orbit_user_space_instrumentation {
+
+// TestProcess forks a new process in the constructor and starts a multi threaded dummy load: A busy
+// loop that spawns and joins threads. The spawned threads perform a busy wait for 15 ms. Four
+// worker threads are kept active. When TestProcess goes out of scope the process is ended.
+class TestProcess {
+ public:
+  TestProcess();
+  ~TestProcess();
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.h:21]: nit: virtual or make class final

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  if (pid_ == 0) {
+    DummyWorkload();
+    exit(0);
+  } else {
+    while (!fs::exists(flag_file_child_started_)) continue;
+  }
+}
+
+TestProcess::~TestProcess() {
+  fs::remove(flag_file_run_child_);
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.cpp:47]: does this variant throw an exception? if so use the one with error_code.

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  flag_file_child_started_ = std::tmpnam(nullptr);
+  Touch(flag_file_run_child_);
+  pid_ = fork();
+  CHECK(pid_ != -1);
+  // Start the workload and have the parent wait for the startup to complete.
+  if (pid_ == 0) {
+    DummyWorkload();
+    exit(0);
+  } else {
+    while (!fs::exists(flag_file_child_started_)) continue;
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.cpp:42]: is this noexcept variant?

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+  // Wait for the traced thread to stop. Timeout after one second.
+  int32_t timeout_ms = 1000;
+  while (true) {
+    int stat_val = 0;
+    const int result_waitpid = waitpid(tid, &stat_val, WNOHANG);
+    if (result_waitpid == -1) {
+      return ErrorMessage(absl::StrFormat(
+          "Wait for thread to get traced failed for tid \"%d\": \"%s\"", tid, strerror(errno)));
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/Attach.cpp:50]: Not important, but: do we want pids/tids enclosed in quotes? Unlike with strings, it doesn't seem too useful.

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+      break;
+    }
+  }
+  return outcome::success();
+}
+
+void DetachAndContinueProcess(pid_t pid) {
+  auto tids = GetTidsOfProcess(pid);
+  for (auto tid : tids) {
+    ptrace(PTRACE_DETACH, tid, nullptr, nullptr);
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/Attach.cpp:103]: Does this always succeed or could there be issues or errors with this?

File: src/UserSpaceInstrumentation/AttachTest.cpp (hunk 7 above)
Comment (dpallotti) [reply src/UserSpaceInstrumentation/AttachTest.cpp:28]: Well Dimitry, I'm glad we finally agree on this. But I have to say that the `absl` version is generally easier to use, `chrono` is sometimes unintuitive. So I'm very much in favor of leaving it as preference.

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 6 above)
Comment (dpallotti) [reply src/UserSpaceInstrumentation/Attach.cpp:72]: We have UNREACHABLE() in `OrbitBase/Logging.h`, it shouldn't cause warnings.

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 1 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 2 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 3 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 4 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 5 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: done

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 8 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp]: It isn't a fake (in my understanding of fake). Just dropped "dummy".

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 9 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp]: as above

File: src/UserSpaceInstrumentation/TestProcess.h (hunk 10 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.h]: done

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 11 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp]: done

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 12 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp]: done

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    } else {
+      return ErrorMessage(
+          absl::StrFormat("PTRACE_ATTACH failed for \"%d\": \"%s\"", tid, strerror(errno)));
+    }
+  }
+  // Wait for the traced thread to stop. Timeout after one second.
+  int32_t timeout_ms = 1000;
+  while (true) {
+    int stat_val = 0;
+    const int result_waitpid = waitpid(tid, &stat_val, WNOHANG);
```
[the remark this answers is not in the document]
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 13 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: Yeah, this makes no sense.

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 14 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp]: Good question. It is somewhere between can't happen and programming error. The errors that might occure *should* be catched in the part where we stop the threads. To play save I expose the return value to the caller - although there is no mitigation for this. I could probably check fail directly as well.

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    // can be looked up in the function 'ptrace_attach' in 'ptrace.c' in the kernel sources.
+    if (errno == ESRCH || errno == EPERM) {
+      return false;
+    }
+    return ErrorMessage(
+        absl::StrFormat("PTRACE_ATTACH failed for %d: \"%s\"", tid, SafeStrerror(errno)));
+  }
+  // Wait for the traced thread to stop. Timeout after one second.
+  int32_t max_attempts = 1000;
+  while (true) {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:45]: If this converted to for (size_t i=0; i < kMaxAttempts; i++) you can drop unreachable below and have lines 65-68 at the end.

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        continue;
+      }
+      OUTCOME_TRY(result, AttachAndStopThread(tid));
+      if (result) {
+        halted_tids.insert(tid);
+      }
+    }
+    // Update the tids; if there are new ones run another round and stop them.
+    process_tids = GetTidsOfProcess(pid);
+    if (process_tids.size() == halted_tids.size()) {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:93]: seems like this could be condition of the while loop.

File: src/UserSpaceInstrumentation/AttachTest.cpp (hunk 7 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/AttachTest.cpp:28]: Hmm, usability in abseil is better. I always need to read up stuff before using std::chrono.

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void> AttachAndStopProcess(pid_t pid) {
+  auto process_tids = GetTidsOfProcess(pid);
+  std::set<pid_t> halted_tids;
+  // Note that the process is still running - it can spawn and end threads at this point.
+  while (true) {
+    for (const auto tid : process_tids) {
+      if (halted_tids.count(tid) == 1) {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:83]: nit: contains?

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  UNREACHABLE();
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void> AttachAndStopProcess(pid_t pid) {
+  auto process_tids = GetTidsOfProcess(pid);
+  std::set<pid_t> halted_tids;
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/Attach.cpp:79]: nit: flat_hash_set?

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+  absl::MutexLock lock{&joinable_threads_mutex_};
+  joinable_threads_.emplace(std::this_thread::get_id());
+}
+
+void TestProcess::Workload() {
+  size_t kNumThreads = 4;
+  std::vector<std::thread> threads;
+  while (fs::exists(flag_file_run_child_) || !threads.empty()) {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.cpp:75]: this can throw an exception - use noexpect variant?

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    // Spawn as many threads as there are missing.
+    while (threads.size() < kNumThreads && fs::exists(flag_file_run_child_)) {
```
Comment (dimitry-) [inline src/UserSpaceInstrumentation/TestProcess.cpp:77]: same here.

**Review by dimitry- — APPROVED** [review]
otherwise lgtm

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 16 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp:45]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 17 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp:93]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 18 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp:83]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 19 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp:79]: done

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 20 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp:75]: done

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 21 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp:77]: done

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+  }
+
+  ptrace(PTRACE_DETACH, tid, nullptr, nullptr);
+  return ErrorMessage(absl::StrFormat("Waiting for the traced thread %d to stop timed out.", tid));
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void> AttachAndStopProcess(pid_t pid) {
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/Attach.cpp:75]: (Is `[[nodiscard]]` needed in the .cpp?)

File: src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/Attach.h [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <sys/types.h>
+
+#include "OrbitBase/Result.h"
+
+namespace orbit_user_space_instrumentation {
+
+[[nodiscard]] ErrorMessageOr<void> AttachAndStopProcess(pid_t pid);
+
+[[nodiscard]] ErrorMessageOr<void> DetachAndContinueProcess(pid_t pid);
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/Attach.h:16]: (This looks very much like one of those cases where `[[nodiscard]]` is added because we add it everywhere but where the caller might want to ignore the result. Not that it matters, but I'm still on the fence with out approach to nodiscard.)

File: src/UserSpaceInstrumentation/Attach.cpp [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+  auto process_tids = GetTidsOfProcess(pid);
+  absl::flat_hash_set<pid_t> halted_tids;
+  // Note that the process is still running - it can spawn and end threads at this point.
+  while (process_tids.size() != halted_tids.size()) {
+    for (const auto tid : process_tids) {
+      if (halted_tids.contains(tid)) {
+        continue;
+      }
+      OUTCOME_TRY(result, AttachAndStopThread(tid));
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/Attach.cpp:84]: Suppose this fails for one thread, aren't all the already stopped thread left in a traced state?

File: src/UserSpaceInstrumentation/TestProcess.cpp [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+  CHECK(!error);
+  int status;
+  waitpid(pid_, &status, 0);
+  CHECK(WIFEXITED(status));
+  fs::remove(flag_file_child_started_, error);
+  CHECK(!error);
+}
+
+void TestProcess::Worker() {
+  constexpr auto kTimeToLiveMs = std::chrono::milliseconds(15);
```
Comment (dpallotti) [inline src/UserSpaceInstrumentation/TestProcess.cpp:61]: I would call this only `kTimeToLive`, as it's not a number in milliseconds but an actual `
std::chrono::duration`.

**Review by dpallotti — APPROVED** [review]
The only real doubt I have is about the `OUTCOME_TRY` in `AttachAndStopProcess`. Other than that it looks good.

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 22 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp:75]: It's enought to put it in the declaration (which makes sense - the compiler usually hasn't seen the definition when issueing a warning).
Removed.

File: src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/Attach.h (hunk 23 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/Attach.h:16]: Hmm, I think enforcing checking for errors here is valid - the underlying calls to ptrace can fail in unusual cases.

File: src/UserSpaceInstrumentation/TestProcess.cpp (hunk 25 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/TestProcess.cpp:61]: done

File: src/UserSpaceInstrumentation/Attach.cpp (hunk 24 above)
Comment (danielfenner) [reply src/UserSpaceInstrumentation/Attach.cpp:84]: Yes, true. There is a similar case in the detach method. Currently I return with an error and leave things in a state that can not be recovered from.

I added code that at least tries to clean up here so that we can recover. However that might go wrong as well. Let's chat briefly offline about how to handle recoverable vs. unrecoverable errors.
